cmd/cloud_sql_proxy/cloud_sql_proxy.go

```diff
@@ -493,5 +496,27 @@ func main() {
 		}),
 		Conns:              connset,
 		RefreshCfgThrottle: refreshCfgThrottle,
-	}).Run(connSrc)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM)
+
+	go func() {
+		<-signals
+		logging.Infof("Received TERM signal. Waiting up to %s seconds before terminating.", *termTimeout)
+
+		termTime := time.Now().Add(*termTimeout)
+		for termTime.After(time.Now()) && proxyClient.ConnectionsCounter > 0 {
+			time.Sleep(1)
+		}
+
+		// Exit cleanly if there are no active connections when we exit
+		if proxyClient.ConnectionsCounter == 0 {
```

Comment:
Shouldn't this and the read in the for loop use atomic.LoadUint64() instead?

Reply:
I don't think it's a good idea to use atomics on a field in a type from another package - it's breaking encapsulation. I would rather an exported method on type Client which appropriately deals with locking/atomics/whatever so that users of the type can't create race conditions.